plugin-base/src/main/kotlin/io/spine/tools/gradle/project/ProjectExts.kt

```diff
@@ -86,8 +87,61 @@ public fun Project.sourceSet(name: SourceSetName): SourceSet = sourceSets.getByN
  * For other source sets, the given source set name would be used as a classifier of the artifact.
  */
 public fun Project.artifact(ssn: SourceSetName): MavenArtifact {
+    val group = group.toString()
+    val artifactName = findMavenPublicationArtifactId() ?: name
+    val version = version.toString()
     val classifier = if (ssn == main) null else ssn.value
-    return MavenArtifact(group.toString(), name, version.toString(), classifier)
+    return MavenArtifact(group, artifactName, version, classifier)
+}
+
+/**
+ * Obtains a parameterless method with the given name.
+ */
+private fun Any.method(name: String): Method? =
+    javaClass.methods.find { m -> m.name == name && m.parameterCount == 0 }
+
+/**
+ * Attempts to obtain the `artifactId` of the first Maven publication in this project.
+ *
+ * Uses reflection to avoid a hard dependency on the Maven Publish plugin classes.
+ * Returns `null` if the `maven-publish` plugin is not applied, if there are no
+ * publications, or if reflection fails for any reason.
+ */
+@Suppress("SwallowedException", "ReturnCount", "CyclomaticComplexMethod")
+private fun Project.findMavenPublicationArtifactId(): String? {
+    return try {
+        // Check if the Maven Publish plugin is applied.
+        if (!pluginManager.hasPlugin("maven-publish")) {
+            return null
+        }
+        // Obtain the "publishing" extension by name to avoid direct API usage.
+        val publishingExt = extensions.findByName("publishing") ?: return null
+
+        // Invoke `getPublications()` reflectively.
+        val getPublications = publishingExt.method("getPublications")
+            ?: return null
+        val publications = getPublications.invoke(publishingExt) ?: return null
+
+        // Helper to extract artifactId from a publication using reflection.
+        fun artifactIdOf(pub: Any?): String? {
+            if (pub == null) return null
+            val method = pub.method("getArtifactId")
+                ?: return null
+            val value = method.invoke(pub) as? String
+            return value?.takeIf { it.isNotBlank() }
+        }
+
+        // Try to treat publications as Iterable; if not, fall back to iterator() via reflection.
+        val iterable = publications as? Iterable<*>
+        if (iterable != null) {
+            return iterable.asSequence().mapNotNull { artifactIdOf(it) }.firstOrNull()
+        }
+
+        null
+    } catch (_: Throwable) {
+        // Be conservative: if anything goes wrong, fall back to the project name.
```

Comment:
I don't think we are falling back to the project name. We return `null`.

Reply:
We return `null` meaning we did not find a Maven publication. As the result, we return `Project.name` in the calling method.